gemd/util/impl.py

```diff
@@ -98,34 +109,118 @@ def _substitute(thing: Any,
     return new
 
 
+def _substitute_inplace(thing: Any,
+                        sub: Callable[[object], object],
+                        applies: Callable[[object], bool],
+                        visited: Dict[object, object] = None) -> object:
+    """
+    Generic recursive in-place substitute function.
+
+    Iteratively crawls the passed structure, substituting elements with sub(element) when
+    applies(element) is true and the element is mutable.
+
+    Parameters
+    ----------
+    thing: Any
+        The object to traverse with substitution.
+    sub: Callable[[object], object]
+        Function which provides substitute for value; should not have side-effects.
+    applies: Callable[[object], bool]
+        Function which defines the domain for the sub function to be invoked.
+
+    """
+    def _key(obj):
+        if obj.__hash__ is not None:
+            return obj
+        else:
+            return id(obj)
+
+    orig_key = _key(thing)
+    if visited is None:
+        visited = {}
+    if orig_key in visited:
+        return visited[orig_key]
+
+    if applies(thing):
+        thing = sub(thing)
+    visited[orig_key] = thing  # Store before we start recursing
+
+    if _cached_isinstance(thing, list):  # Change list in place
+        for i, x in enumerate(thing):
+            thing[i] = _substitute_inplace(x, sub, applies, visited)
+    elif _cached_isinstance(thing, tuple):  # Tuples are immutable; regenerate
+        thing = tuple(_substitute_inplace(x, sub, applies, visited) for x in thing)
+        visited[orig_key] = thing  # We mutated it
+    elif _cached_isinstance(thing, dict):  # Change dict in place, both keys & values
+        remove = set()  # Store todos because can't mutate a dict in a loop
+        update = dict()
+        for k, v in thing.items():
+            new_k = _substitute_inplace(k, sub, applies, visited)
+            new_v = _substitute_inplace(v, sub, applies, visited)
+            if id(k) != id(new_k):
+                remove.add(k)
+                update[new_k] = v
+            if id(v) != id(new_v):
+                update[new_k] = new_v
+        for k in remove:
+            thing.pop(k, None)
+        thing.update(update)
+    elif _cached_isinstance(thing, DictSerializable):
+        for k, v in thing.as_dict().items():  # Assume key can't change b/c it's an attribute
+            new_v = _substitute_inplace(v, sub, applies, visited)
+            if id(v) != id(new_v):
+                _setter_by_name(type(thing), k)(thing, new_v)
+
+    return thing
+
+
+@functools.lru_cache(maxsize=None)
+def _setter_by_name(clazz: type, name: str) -> Callable:
+    """Internal method to get the setter method for an attribute."""
+    from gemd.entity.object import IngredientRun
+
+    def _emulator(inner_name: str) -> Callable:
+        return lambda self, value: setattr(self, inner_name, value)
+
+    prop = next((x[1] for x in getmembers(clazz) if x[0] == name), None)
```

Comment:
I'm trying to parse what this method does and am having a really tough time doing so.

It seems that since `IngredientRun` has a separate _name method, you need to generate a setter for calls in the above methods?

Couldn't we just change `IngredientRun` to conform and use some syntactic sugar to be non-breaking? I think this may be too clever for where there seems like an easier fix by just changing IngredientRun.

Reply:
Ooof.

`BaseEntity` has a name property with a setter and getter.  Since you are not supposed to set `IngredientRun.name` directly, the method is overridden with a method that raises an `AttributeError`.  This is further complicated by the fact that the object will contain a `_name` value when being retrieved from the server or after a link to an `IngredientSpec` is broken, since in those cases it has a name, it just can't be accessed dynamically.

We might be able to achieve the situation described by editing the `property` in the class members, but I recall having difficult implementing that when this first raised its head.  If we could get past that, this could be handled with:
```
    elif prop.fset is None:
        setter = _emulator(f"_name")
```

Reply:
Addendum: I just quickly looked into ^^.  Because the `__init__` method of `BaseAttribute` calls the name setter.  `name` is also a required argument of `BaseAttribute`.  It's functionally resolvable if we modify that to 
```
        if not isinstance(self, IngredientRun):
            self.name = name
```
but a parent class inheriting from a child is pretty hideous as well.

Reply:
This particular block of code  is what I'm hung up on. I don't want to block this PR on it, but I would urge that this be followed up on/ticketed as a piece of maintenance work. It seems like that minor issue is necessitating pretty confusing/hacky workarounds in a case like this, and that will continue to be a hindrance if not sorted out.